PLANET 3757  Happy Point block

sagarsdeshmukh opened the pull request.

kirdia commented on classes/blocks/class-happypoint.php:
```diff
@@ -0,0 +1,121 @@
+/**
+ * Class Happypoint_Controller
+ *
+ * @package P4BKS\Controllers\Blocks
```
Wrong package here

kirdia commented on classes/blocks/class-happypoint.php:
```diff
@@ -0,0 +1,121 @@
+	 *
+	 * @const string BLOCK_NAME.
+	 */
+	const BLOCK_NAME = 'Happypoint';
```
better use the old block name (and lowercase chars) for consistency

kirdia commented on classes/blocks/class-happypoint.php:
```diff
@@ -0,0 +1,121 @@
+	}
+
+
+	public function __construct() {
```
Missing comment here

kirdia commented on classes/blocks/class-happypoint.php:
```diff
@@ -0,0 +1,121 @@
+						'default' => '',
+					],
+					'opacity'             => [
+						'type'    => 'string',
```
this should be a number

kirdia commented on react-blocks/src/blocks/Happypoint/Happypoint.js:
```diff
@@ -0,0 +1,138 @@
+          </div>
+
+          <div>
+            <TextControl
```
it could be of type 'number'

kirdia commented on react-blocks/src/blocks/Happypoint/Happypoint.js:
```diff
@@ -0,0 +1,138 @@
+export class Happypoint extends Component {
+    constructor(props) {
+      super(props);
+      this.state = {
```
setting state not necessary, here could be removed

kirdia commented on react-blocks/src/styles/blocks/Happypoint.scss:
```diff
@@ -0,0 +1,9 @@
+#editor .editor-writing-flow > div > div > div > .editor-block-list__layout > div[data-type="planet4-blocks/happypoint"] {
```
please remove these to keep blocks layout consistent on the editor.
the second rule is not used.

kirdia commented on react-blocks/src/blocks/Happypoint/Happypoint.js:
```diff
@@ -0,0 +1,109 @@
+                onChange={this.props.onFocalPointChange}
+              />
+
+              {/*<img src={url} alt={alt} />*/}
```
please remove comment here

kirdia commented on react-blocks/src/blocks/Happypoint/Happypoint.js:
```diff
@@ -0,0 +1,109 @@
+          </InspectorControls>
+          <div>
+            {url ? (
+            <>
```
please use Fragment here also

kirdia commented:
please remove these also from commit
react-blocks/build/style.js.map
react-blocks/build/editorStyle.js.map

sagarsdeshmukh commented:
@kirdia - I amend the last commit. please have a look.

pablocubico commented:
@sagarsdeshmukh This looks great! It's only missing a few pieces (according to the old happy point block):

- A title for the focal point picker, it used to say: "Select focus point for background image"
- And I guess these fields are missing (Iframe URL, background opacity, etc): 

![image](https://user-images.githubusercontent.com/340766/63402338-5db0b400-c3b1-11e9-83b1-70b021f532a8.png)

sagarsdeshmukh commented:
Hi @pablocubico , Thanks. I added a label on focal point picker component.
Regarding the Iframe URL, background opacity fields, those are available in right side block settings.
![image](https://user-images.githubusercontent.com/5357471/63405243-b9a42880-c403-11e9-82a8-1b36095ef4e5.png)

pablocubico commented:
@sagarsdeshmukh ohhh, I didn't notice the sidebar! ignore me then. 

Please check the `background-position` versus `object-position` issue, it's in this note: https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/3/files#diff-ced48b9df02dad6ef33db067fcf0cc76R6

If you test it in the frontend, you'll see the focal point does not change with `background-position`, since the image sits on an `img` tag and not a `background: url(...)` style, so I think you'll have to change it to `object-position`, I added a sample in the comment.

sagarsdeshmukh commented:
> @sagarsdeshmukh ohhh, I didn't notice the sidebar! ignore me then.
> 
> Please check the `background-position` versus `object-position` issue, it's in this note: https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/3/files#diff-ced48b9df02dad6ef33db067fcf0cc76R6
> 
> If you test it in the frontend, you'll see the focal point does not change with `background-position`, since the image sits on an `img` tag and not a `background: url(...)` style, so I think you'll have to change it to `object-position`, I added a sample in the comment.

The `object-position` works. Thanks 👍 I updated the PR, please have a look.

pablocubico commented:
Hey @sagarsdeshmukh !  I just noticed the styles for the frontend side are not loading, I think you need to:
- copy the styles from the old Blocks plugin to this repo, as the other ones, for example, check how the Articles block's styles are included: https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/blob/develop/react-blocks/src/styles/style.scss#L7
-  then wait for this to be merged: https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/17
- this block would also need the `blocks_wide` script which expands the container the whole screen width, I had to add it in the EN Gutenberg repo, check it out: https://github.com/greenpeace/planet4-plugin-gutenberg-engagingnetworks/blob/master/classes/class-loader.php#L399 , maybe you can copy the `blocks_wide` script into this repo and do the same 🤔 


And one last thing, I noticed all the original Blocks styles are loaded in the editor:
![image](https://user-images.githubusercontent.com/340766/63661957-ffaa1500-c792-11e9-8599-d10739ed9b6b.png)

But I couldn't find where the file is added... Is it me? What could I be doing wrong? pinging @comzeradd too for this one.

sagarsdeshmukh commented:
Hi @pablocubico ,
I added the frontend styles for HP block and `blocks_wide` js script.
Along with these I added `edit image` and `remove image` controls in `block control` panel, replace the `MediaPlaceholder` component with `MediaUpload`. 
I checked the block styles loading in editor issue, but I didn't able to reproduce it. could you please 
 confirm on it.

pablocubico commented:
Hey @sagarsdeshmukh I'm not seeing the blocks styles now, it may had something to do with my environment. 🤔 

Testing again: I disabled all P4 plugins except for the Gutenberg blocks now, just to make sure is not loading any styles from other places; I also updated the branch locally to the latest `develop` branch. Then I got this full-width background:

<img width="848" alt="Captura de pantalla 2019-08-28 a la(s) 23 25 52" src="https://user-images.githubusercontent.com/340766/63907687-ca036700-c9f2-11e9-8841-a5e715f8423d.png">

Do you get the same? It can be fixed by adding this to `happypoint.scss`:
```diff
 .happy-point-block-wrap {
   @include background-before-opacity($light-blue-bg);
   width: 100vw;
+  body.wp-admin & {
+    width: 100%;
+  }
```

Which makes the width `100%` instead of `100vw` when inside the editor.

The edit & remove feature looks great! For the icon placement I would suggest placing them next to the image, because of a "proximity" Gestalt principle (here is [a nice article about it](https://medium.muz.li/gestalt-principles-in-ui-design-6b75a41e9965)), but I wouldn't change it now because the design team will probably review the UX for all the blocks when we finish porting them.

Sorry, I made it too long, in brief, I think you need to:
- Update the branch to latest develop
- Add the `.wp-admin` specific rule
- Fix the URL for the `blocks_wide` script (see comment in the block's PHP class)

Great work, keep those blocks coming! 🥊

sagarsdeshmukh commented:
> Hey @sagarsdeshmukh I'm not seeing the blocks styles now, it may had something to do with my environment. 🤔
> 
> Testing again: I disabled all P4 plugins except for the Gutenberg blocks now, just to make sure is not loading any styles from other places; I also updated the branch locally to the latest `develop` branch. Then I got this full-width background:
> 
> <img alt="Captura de pantalla 2019-08-28 a la(s) 23 25 52" width="848" src="https://user-images.githubusercontent.com/340766/63907687-ca036700-c9f2-11e9-8841-a5e715f8423d.png">
> 
> Do you get the same? It can be fixed by adding this to `happypoint.scss`:
> 
> ```diff
>  .happy-point-block-wrap {
>    @include background-before-opacity($light-blue-bg);
>    width: 100vw;
> +  body.wp-admin & {
> +    width: 100%;
> +  }
> ```
> 
> Which makes the width `100%` instead of `100vw` when inside the editor.
> 
> The edit & remove feature looks great! For the icon placement I would suggest placing them next to the image, because of a "proximity" Gestalt principle (here is [a nice article about it](https://medium.muz.li/gestalt-principles-in-ui-design-6b75a41e9965)), but I wouldn't change it now because the design team will probably review the UX for all the blocks when we finish porting them.
> 
> Sorry, I made it too long, in brief, I think you need to:
> 
> * Update the branch to latest develop
> * Add the `.wp-admin` specific rule
> * Fix the URL for the `blocks_wide` script (see comment in the block's PHP class)
> 
> Great work, keep those blocks coming! 🥊

Thanks @pablocubico , I updated the PR. Please have a look.
